feat: Add APIs page for JavaScript (#12605)
getsentry/sentry-docs: docs/platforms/javascript/common/apis.mdx

Reviewer: I think the styling could still be improved a bit. Something about this feels misaligned or empty (?):

<img width="747" alt="Screenshot 2025-03-28 at 12 00 17" src="https://github.com/user-attachments/assets/8f192ccc-286d-43ee-a8f2-f8b41002cb4f" />

Like the title of the expandable not having any separator to the content and the heading within the content being larger than the title of the expandable.

We can do that in a follow-up PR tho.

Author: yeah, I agree this can be prettier/nicer for sure!

Reviewer 2: Maybe it's because all margins have a different width 🤔 
Created a comment about this for the "examples" (but it's the same in "Parameters"): https://github.com/getsentry/sentry-docs/pull/12605#discussion_r2018656467

Reviewer 3: General question: will this page have a higher relevance for search? For example, when I look for "addIntegration", it would be great to have this page listed in the results somewhere near the top

        { name: 'op', type: 'string', description: 'The operation name for the span. This is used to group spans in the UI' },
        { name: 'forceTransaction', type: 'boolean', description: 'If true, the span will be forced to be sent as a transaction, even if it is not the root span.' },
        { name: 'parentSpan', type: 'Span | null', description: 'The parent span for the new span. If not provided, the current span will be used.' },
        { name: 'onlyIfParent', type: 'boolean', description: 'If true, the span will only be created if there is an active span.' },
      ]
    },
  },
  { name: "callback", type: "(span: Span) => T", required: true },
]}>
  Starts a new span, that is active in the provided callback.
  This span will be a child of the currently active span, if there is one.

Any spans created inside of the callback will be children of this span.

The started span will automatically be ended when the callback returns, and will thus measure the duration of the callback. The callback cann also be an async function.

  <Expandable title='Examples'>
  ```javascript
  // Synchronous example
 Sentry.startSpan({ name: 'my-span' }, (span) => {
    measureThis();
  });

> // Asynchronous example

Author: this is probably outside of the scope of this PR then, as this is just the regular `<Expandable>` usage - imho the margins here are _generally_ fine 🤔 maybe it just looks a bit weird because there is no text, I'll add a sentence above the code block, maybe that helps!

> // Asynchronous example

Author: hmm, I'd say that does not really make it much better xD
![image](https://github.com/user-attachments/assets/2dc78fc4-e515-4eb8-9590-0c6c2e652c55)

> // Asynchronous example

Reviewer 2: It's okay to keep it like this in this PR :D
But it makes sense to check the margins in a follow-up.

> // Asynchronous example

Author: for what it's worth, this should be a bit better now with the changes to the expandable!

---

> title: APIs

Reviewer: Another random thing I spotted: The right sidebar is so large that it goes beyond the screen, however when the "current heading" updates while scrolling the highlighted heading doesn't scroll into view.

                  type: "Record<string, unknown>",
                  description:
                    "Additional data that should be sent with the exception.",
                },
                {
                  name: "tags",
                  type: "Record<string, string>",
                  description:
                    "Additional tags that should be sent with the exception.",
                },
                {
                  name: "contexts",
                  type: "Record<string, Record<string, unknown>>",
                },
                { name: "fingerprint", type: "string[]" },
              ],
            },
            description:
              "Optional additional data to attach to the Sentry event.",
          },
        ],
      },
      description:
        "Optional hint object containing additional information about the feedback.",
    },
  ]}
>
  Send user feedback to Sentry.
</SdkApi>

<SdkApi
  name="getFeedback"
  signature="function getFeedback(): ReturnType<feedbackIntegration> | undefined"
>
  Get the feedback integration, if it has been added. This can be used to access
  the feedback integration in a type-safe way.
</SdkApi>

> <SdkApi

Reviewer: Something's busted here. It doesn't render past sendFeedback.

Author: huh, pushed a fix hopefully!

          { name: "recoveryThreshold", type: "number" },
        ],
      },
    },
  ]}
>
  Create a cron monitor check in and send it to Sentry.
</SdkApi>

<SdkApi
name="withMonitor"
signature={`function withMonitor(
monitorSlug: string,
callback: () => any,
monitorConfig?: MonitorConfig
): string`}
categorySupported={['server']}
parameters={[
{
  name: "monitorSlug",
  type: 'string',
  required: true,
},
 {
  name: "callback",
  type: '() => any',
  required: true,
},
{
  name: "monitorConfig",
  type: {
    name: "MonitorConfig",
    properties: [
      {
        name: "schedule",
        type: '{ type: "crontab", value: string } | { type: "interval", value: number, unit: "year" | "month" | "day" | "hour" | "minute" }',
        required: true,
      },
      { name: "checkinMargin", type: "number" },
      { name: "maxRuntime", type: "number" },
      { name: "timezone", type: "string" },
      { name: "failureIssueThreshold", type: "number" },
      { name: "recoveryThreshold", type: "number" },
    ],
  },
},
]}
>
Wraps a callback with a cron monitor check in. The check in will be sent to Sentry when the callback finishes.
</SdkApi>
</PlatformCategorySection>
```

> ````

Reviewer: This looks like odd syntax.

---
title: APIs
description: "Learn more about APIs of the SDK."
customCanonicalTag: "/platforms/javascript/apis/"
sidebar_order: 3
---

> This page shows all available to-level APIs of the SDK. You can use them like this:

Reviewer 3: We could add a bit more information here at the top (e.g., very briefly explain the purpose of these APIs).
It could also be good to use the word "function" to make it clear that these are SDK APIs and not REST APIs (something like "... the primary functions provided by the SDK for interacting with ...").

Let me know if I can support you with this :)

Author: good point, will add a bit more context here!

```javascript
import * as Sentry from "@sentry/browser";

Sentry.setTag("tag", "value");
```

> ## Available Options

Reviewer 3: Should this be "Available APIs"?

Author: yes 😅

> <TableOfContents ignoreIds={["available-options"]} />

Reviewer 3: Just a note, probably not a topic right now:
When you open the page you see two ToC's - one in the center and one to the right.
I can imagine that users gravitate towards the ToC in the center to find the API they're looking for (it is more eye-catching than the ToC in the sidebar). So, it is definitely valuable and needs to be here!
Still, I wish we could toggle the ToC (visible by default); especially considering that the list will become longer rather than shorter in the future 😅

Author: hmm, I get what you mean. What would you propose (for a follow up):

1. allow to toggle the TOC in the main page?
2. Do not show the TOC in the sidebar?
3. Something else...?

Reviewer 3: My proposition:
3. allow to toggle the second level of list elements in the main area (default: hidden)
 
  Core APIs >
  Capturing Events >
  Enriching Events v
      setTag
      setTags
      
I understand if this is too extra :D

## Core APIs

<SdkApi
  name="init"
  signature="function init(options: InitOptions): Client | undefined"
>
  Initialize the SDK with the given options. See{" "}
  <PlatformLink to="/configuration/options/">Options</PlatformLink> for the
  options you can pass to `init`.
</SdkApi>

<SdkApi name="getClient" signature="function getClient(): Client | undefined">
  Returns the currently active client.
</SdkApi>

<SdkApi
  name="setCurrentClient"
  signature="function setCurrentClient(client: Client): void"
>
  Make the given client the current client. You do not need this if you use

>   `init()`, this is only necessary if you are manually setting up a client.˝

Reviewer 4: ```suggestion
  `init()`, this is only necessary if you are manually setting up a client.
```

</SdkApi>

<SdkApi
  name="lastEventId"
  signature="function lastEventId(): string | undefined"
>
  Returns the ID of the last sent error event. Note that this does not guarantee
  that this event ID exists, as it may have been dropped along the way.
</SdkApi>

<SdkApi
  name="flush"
  signature="function flush(timeout?: number): Promise<boolean>"
  parameters={[
    {
      name: "timeout",
      type: "number",
      description:
        "Maximum time in ms the client should wait to flush its event queue. Omitting this parameter will cause the client to wait until all events are sent before resolving the promise.",
    },
  ]}
>
  Flushes all pending events.
</SdkApi>

<SdkApi name="isEnabled" signature="function isEnabled(): boolean">
  Returns true if the SDK is initialized & enabled.
</SdkApi>

<SdkApi
  name="close"

>   signature="function flush(timeout?: number): Promise<boolean>"

Reviewer 4: ```suggestion
  signature="function close(timeout?: number): Promise<boolean>"
```

Reviewer 4: Also, should we put this directly under `flush` since the two are related?

  parameters={[
    {
      name: "timeout",
      type: "number",
      description:
        "Maximum time in ms the client should wait to flush its event queue. Omitting this parameter will cause the client to wait until all events are sent before resolving the promise.",
    },
  ]}
>
  Flushes all pending events and disables the SDK. Note that this does not
  remove any listeners the SDK may have set up.
</SdkApi>

<SdkApi
  name="addEventProcessor"
  signature="function addEventProcessor(processor: EventProcessor): void"
  parameters={[
    {
      name: "processor",
      type: "(event: Event, hint: EventHint) => Event | null | Promise<Event | null>",
    },
  ]}
>
  Adds an event processor to the SDK. An event processor receives every event
  before it is sent to Sentry. It can either mutate the event (and return it) or
  return `null` to discard the event. Event processors can also return a
  promise, but it is recommended to use this only when necessary as it slows
  down event processing.

> </SdkApi>

Reviewer 4: IIRC we discussed removing some smaller dedicated pages in favour of this API page. I'm not sure if we're still going to do this. In case of `addEventProcessor`, should we link to the [event processors](https://docs.sentry.io/platforms/javascript/enriching-events/event-processors/) page?

Author: I think for this case, we can/should eventually remove the event processors page 🤔 But I'll port some more of it's content in here!

Reviewer 4: also fine with me!

<SdkApi
  name="addIntegration"
  signature="function addIntegration(integration: Integration): void"
>
  Adds an integration to the SDK. This can be used to conditionally add
  integrations after `Sentry.init()` has been called. Note that it is
  recommended to pass integrations to `init` instead of calling this method,
  where possible.

> </SdkApi>

Reviewer 4: same q here - should we link to the [Integrations](https://docs.sentry.io/platforms/javascript/configuration/integrations/) page?

<SdkApi name="lazyLoadIntegration" signature="function lazyLoadIntegration(name: string, scriptNonce?: string): Promise<Integration>" categorySupported={['browser']}>
  Lazy load an integration. This expects the name to be e.g. `replayIntegration`. It will load the script from the CDN, and return a promise that resolves to the integration, which can then be added to the SDK using `addIntegration`:

```javascript
Sentry.lazyLoadIntegration("replayIntegration")
  .then((integration) => {
    Sentry.addIntegration(integration);
  })
  .catch((error) => {
    // Make sure to handle errors here!
    // This rejects e.g. if the CDN bundle cannot be loaded
  });

> ```

Reviewer 4: same q here

If you use a bundler, using e.g. `const { replayIntegration } = await import('@sentry/browser')` is recommended instead.

</SdkApi>

> ## Scopes
> 
> <SdkApi name="getCurrentScope" signature="function getCurrentScope(): Scope">
>   Returns the current scope. See{" "}
>   <PlatformLink to="/enriching-events/scopes/">Scopes</PlatformLink> for more
>   information.

Reviewer 4: l: I think for this section we can link to the Scopes page directly below the heading, wdyt? If so, we can remove the individual links or we target the links to the specific sections on the Scopes page.

```suggestion
## Scopes

See{" "}
  <PlatformLink to="/enriching-events/scopes/">Scopes</PlatformLink> for more
  information.

<SdkApi name="getCurrentScope" signature="function getCurrentScope(): Scope">
  Returns the current scope. 
```

Note that in most cases you should not use this API, but instead use `withScope` to generate and access a local scope. There are no guarantees about the consistency of `getCurrentScope` across different parts of your application, as scope forking may happen under the hood at various points.

</SdkApi>

<SdkApi
  name="withScope"
  signature="function withScope(callback: (scope: Scope) => void): void"
>
  Forks the current scope and calls the callback with the forked scope. See{" "}
  <PlatformLink to="/enriching-events/scopes/">Scopes</PlatformLink> for more
  information.
</SdkApi>

<SdkApi
  name="getIsolationScope"
  signature="function getIsolationScope(): Scope"
>
  Returns the current isolation scope.
</SdkApi>

<SdkApi
  name="withIsolationScope"
  signature="function withIsolationScope(callback: (scope: Scope) => void): void"
>

>   Forks the current isolation scope and calls the callback with the forked

Reviewer 4: super-l: Not sure if "forking" is as widely understood as "cloning" but happy to leave this up to you

```suggestion
  Clones the current isolation scope and calls the callback with the cloned
```

Reviewer 4: sounds good!

>   scope. See <PlatformLink to="/enriching-events/scopes/">Scopes</PlatformLink>{" "}

Reviewer 4: l: If we go with my suggestion above, I'd remove the link here or link to the isolation scope section

```suggestion
  scope.
```

  for more information.
</SdkApi>

<SdkApi name="getGlobalScope" signature="function getGlobalScope(): Scope">

>   Returns the global scope. See{" "}
>   <PlatformLink to="/enriching-events/scopes/">Scopes</PlatformLink> for more
>   information.

Reviewer 4: l: same here OR alternatively link directly to the global scope section on the scope page?
```suggestion
  Returns the global scope.
```

</SdkApi>

## Capturing Events

<SdkApi
  name="captureException"
  signature={`function captureException(
  exception: unknown,
  captureContext?: CaptureContext
): EventId`}
  parameters={[
    {
      name: "exception",
      required: true,
      type: "unknown",

>       description: "The exception to capture. This is usually an Error object.",

Reviewer 4: super-l: Maybe rephrase this a bit to explain `unknown`?
```suggestion
      description: "The exception to capture. For best results, pass an `Error` object but it accepts any kind of value.",
```

              { name: "ip_address", type: "string" },
              { name: "username", type: "string" },
            ],
          },
        },
        {
          name: "level",
          type: '"fatal" | "error" | "warning" | "log" | "info" | "debug"',
        },
        {
          name: "extra",
          type: "Record<string, unknown>",
          description:
            "Additional data that should be sent with the exception.",
        },
        {
          name: "tags",
          type: "Record<string, string>",
          description:
            "Additional tags that should be sent with the exception.",
        },
        { name: "contexts", type: "Record<string, Record<string, unknown>>" },
        { name: "fingerprint", type: "string[]" },
      ],
    },
    description: "Optional additional data to attach to the Sentry event.",
  },
]}
>
Capture a message event and send it to Sentry. Optionally, instead of a
`CaptureContext`, you can also pass a `SeverityLevel` as second argument, e.g.
`"error"` or `"warning"`.

Messages show up as issues on your issue stream, with the message as the issue name.

</SdkApi>

## Enriching Events

> <SdkApi

Reviewer 4: l: should we link to Enriching Events here? Similarly to my comment in scopes, we can also just link to specific sections in the respective methods

Author: this I would like to mostly get rid of eventually - most of the `enrichting events" pages are really just describing a single API or so!

        { name: 'op', type: 'string', description: 'The operation name for the span. This is used to group spans in the UI' },
        { name: 'forceTransaction', type: 'boolean', description: 'If true, the span will be forced to be sent as a transaction, even if it is not the root span.' },
        { name: 'parentSpan', type: 'Span | null', description: 'The parent span for the new span. If not provided, the current span will be used.' },
        { name: 'onlyIfParent', type: 'boolean', description: 'If true, the span will only be created if there is an active span.' },
      ]
    },
  },
  { name: "callback", type: "(span: Span) => T", required: true },
]}>
  Starts a new span, that is active in the provided callback.
  This span will be a child of the currently active span, if there is one.

Any spans created inside of the callback will be children of this span.

The started span will automatically be ended when the callback returns, and will thus measure the duration of the callback. The callback cann also be an async function.

>   <Expandable title='Examples'>

Reviewer 3: I like how you've used the Expandable component throughout the page – it makes the site look very neat! The consistency in use is also great and helps users find their way around 👍

Reviewer 3: Question: How do you currently decide which function to add an example to?

I'm just wondering if it would make sense to add more? I understand that some will not really need one, though

Author: really very much a gut-feeling thing, where I feel the API is more complex/harder to understand without an example 😅 We can def. add more where we feel like it!

  if you opt-out of navigation spans, you can use this method to manually start
  such a span. Please note that this function will do nothing if
  `browserTracingIntegration` has not been enabled.
</SdkApi>

<SdkApi
  name="withActiveSpan"
  signature="function withActiveSpan<T>(span: Span | null, callback: () => T): T"
>
  Runs the provided callback with the given span as the active span. If `null`
  is provided, the callback will have no active span.
</SdkApi>

## Tracing Utilities

These utilities can be used for more advanced tracing use cases.

<SdkApi name="spanToJSON" signature="function spanToJSON(span: Span): SpanJSON">
  Convert a span to a JSON object.
</SdkApi>

<SdkApi
  name="updateSpanName"
  signature="function updateSpanName(span: Span, name: string): void"
>
  Update the name of a span. Use this over `span.updateName(name)` to ensure
  that the span is updated in all backends.
</SdkApi>

<SdkApi
  name="getActiveSpan"
  signature="function getActiveSpan(): Span | undefined"
>
  Get the currently active span.
</SdkApi>

<SdkApi name="getRootSpan" signature="function getRootSpan(span: Span): Span">
  Get the root span of a span.
</SdkApi>

## Sessions

> <SdkApi name="startSession" signature="function startSession(): void">

Reviewer 4: m: I would add a sentence about the session and release health here to avoid confusion with what users would usually think a session is.

Author: lol, this shows: we do not have any page in docs about sessions, I think? 😬 This is the closest we have, I guess it's ok... http://localhost:3000/platforms/javascript/guides/react/configuration/releases/#sessions
